Entities/AtomGroup.cpp

```diff
@@ -1,2618 +1,1750 @@
-	}
+		// If all Atoms are intersecting, we're screwed?!
+		if (intersectingAtoms.size() >= m_Atoms.size()) {
+			return false;
+		}
+
+		Vector exitDirection = Vector();
+
+		// Go through all intersecting Atoms and find their average inverse normal
+		for (const Atom *intersectingAtom : intersectingAtoms) {
+			exitDirection += m_OwnerMO->RotateOffset(intersectingAtom->GetNormal());
+		}
+
+		// We don't have a direction to go, so quit
+		// TODO: Maybe use previous position to create an exit direction instead then?
+		if (exitDirection.IsZero()) {
+			return false;
+		}
+
+		// Invert and set appropriate length
+		exitDirection = -exitDirection;
+		exitDirection.SetMagnitude(m_OwnerMO->GetDiameter());
+
+		// See which of the intersecting Atoms has the longest to travel along the exit direction before it clears
+		float longestDistance = 0.0F;
+
+		Vector clearPos = Vector();
+		Vector atomExitVector = Vector();
+		Vector totalExitVector = Vector();
 
-    // Tell both MO's that they have hit an MO, and see if they want to continue
-	if (m_pOwnerMO->OnMOHit(pIntersectedMO)) {
-		return false;
+		for (const Atom *intersectingAtom : intersectingAtoms) {
+			atomPos = intersectingAtom->GetCurrentPos();
+
+			if (strengthThreshold <= 0.0F) {
+				rayHit = g_SceneMan.CastMaterialRay(atomPos, exitDirection, g_MaterialAir, clearPos, 0, false);
+			} else {
+				rayHit = g_SceneMan.CastWeaknessRay(atomPos, exitDirection, strengthThreshold, clearPos, 0, false);
+			}
+
+			if (rayHit) {
+				// Determine the longest clearing distance so far
+				atomExitVector = clearPos - atomPos;
+				if (atomExitVector.GetMagnitude() > longestDistance) {
+					// We found the Atom with the longest to travel along the exit direction to clear, so that's the distance to move the whole object to clear all its Atoms.
+					longestDistance = atomExitVector.GetMagnitude();
+					totalExitVector = atomExitVector;
+				}
+			}
+		}
+
+		// If the exit vector is too large, then avoid the jarring jump and report that we didn't make it out
+		if (totalExitVector.GetMagnitude() > m_OwnerMO->GetRadius()) {
+			//position += totalExitVector / 2;
+			return false;
+		}
+
+		// Now actually apply the exit vectors to this 
+		position += totalExitVector;
+
+		// TODO: this isn't really true since we don't check for clearness after moving the position
+		return true;
 	}
-	if (pIntersectedMO->OnMOHit(m_pOwnerMO->GetRootParent())) {
-		return false;
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	bool AtomGroup::ResolveMOSIntersection(Vector &position) {
+		if (!m_OwnerMO->m_HitsMOs) {
+			return true;
+		}
+
+		MovableObject *intersectedMO = nullptr;
+		MOID currentMOID = g_NoMOID;
+		MOID hitMOID = g_NoMOID;
+
+		Vector atomOffset = Vector();
+		Vector atomPos = Vector();
+		Vector atomNormal = Vector();
+
+		// First go through all Atoms to find the first intersection and get the intersected MO
```

Comment:
A lot of my comments in ResolveTerrainIntersection apply to this method. Maybe it's possible to even reuse some code, but I can't say for sure without more inspection and comparison.